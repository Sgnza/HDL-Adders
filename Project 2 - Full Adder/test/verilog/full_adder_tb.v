`timescale 1ps/1ps
//`include "../../../Project 1 - Half Adder/src/verilog/half_adder.v"
`include "../../src/verilog/full_adder.v"
module full_adder_tb;

reg a_tb;
reg b_tb;
reg cin_tb;
wire sum_tb;
wire cout_tb;

full_adder fa0(
    .a(a_tb),
    .b(b_tb),
    .cin(cin_tb),
    .sum(sum_tb),
    .cout(cout_tb)
);

task verify;
    input a_task;
    input b_task;
    input cin_task;
    input sum_task;
    input cout_task;

    begin
        a_tb = a_task;
        b_tb = b_task;
        cin_tb = cin_task;
        #1;
        if(sum_task == sum_tb && cout_task == cout_tb) begin
            $display("CORRECTO: Dato de entrada a: %b, Dato de entrada b: %b, Dato de entrada cin: %b, Dato de salida s esperado: %b, Dato de salida s obtenido: %b, Dato de salida c esperado: %b, Dato de salida c obtenido: %b", a_tb, b_tb, cin_task, sum_task, sum_tb, cout_task, cout_tb);
        end else begin
            $display("INCORRECTO: Dato de entrada a: %b, Dato de entrada b: %b, Dato de entrada cin: %b, Dato de salida s esperado: %b, Dato de salida s obtenido: %b, Dato de salida c esperado: %b, Dato de salida c obtenido: %b", a_tb, b_tb, cin_task, sum_task, sum_tb, cout_task, cout_tb);

        end
    end

endtask

initial begin
  $dumpfile("full_adder_tb.vcd");
  $dumpvars(0, full_adder_tb);
end

initial begin

verify(
    1'b0,
    1'b0,
    1'b0,
    1'b0,
    1'b0
);

verify(
    1'b0,
    1'b0,
    1'b1,
    1'b1,
    1'b0
);

verify(
    1'b0,
    1'b1,
    1'b0,
    1'b1,
    1'b0
);

verify(
    1'b0,
    1'b1,
    1'b1,
    1'b0,
    1'b1
);

verify(
    1'b1,
    1'b0,
    1'b0,
    1'b1,
    1'b0
);

verify(
    1'b1,
    1'b0,
    1'b1,
    1'b0,
    1'b1
);

verify(
    1'b1,
    1'b1,
    1'b0,
    1'b0,
    1'b1
);

verify(
    1'b1,
    1'b1,
    1'b1,
    1'b1,
    1'b1
);

end

endmodule