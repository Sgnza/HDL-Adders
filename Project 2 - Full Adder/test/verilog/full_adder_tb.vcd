$date
	Sat Jan 13 10:46:46 2024
$end
$version
	Icarus Verilog
$end
$timescale
	1ps
$end
$scope module full_adder_tb $end
$var wire 1 ! sum_tb $end
$var wire 1 " cout_tb $end
$var reg 1 # a_tb $end
$var reg 1 $ b_tb $end
$var reg 1 % cin_tb $end
$scope module fa0 $end
$var wire 1 # a $end
$var wire 1 $ b $end
$var wire 1 % cin $end
$var wire 1 " cout $end
$var wire 1 ! sum $end
$var wire 1 & ha1_cout $end
$var wire 1 ' ha0_sum $end
$var wire 1 ( ha0_cout $end
$scope module ha0 $end
$var wire 1 # a $end
$var wire 1 $ b $end
$var wire 1 ' sum $end
$var wire 1 ( cout $end
$upscope $end
$scope module hf1 $end
$var wire 1 % a $end
$var wire 1 ' b $end
$var wire 1 ! sum $end
$var wire 1 & cout $end
$upscope $end
$upscope $end
$scope task verify $end
$var reg 1 ) a_task $end
$var reg 1 * b_task $end
$var reg 1 + cin_task $end
$var reg 1 , cout_task $end
$var reg 1 - sum_task $end
$upscope $end
$upscope $end
$enddefinitions $end
$comment Show the parameter values. $end
$dumpall
$end
#0
$dumpvars
0-
0,
0+
0*
0)
0(
0'
0&
0%
0$
0#
0"
0!
$end
#1
1!
1%
1-
1+
#2
1'
0%
1$
0+
1*
#3
0!
1"
1&
1%
1,
0-
1+
#4
1!
0"
0&
0%
0$
0(
1#
0,
1-
0+
0*
1)
#5
0!
1"
1&
1%
1,
0-
1+
#6
0'
0&
0%
1(
1$
0+
1*
#7
1!
1%
1-
1+
#8
