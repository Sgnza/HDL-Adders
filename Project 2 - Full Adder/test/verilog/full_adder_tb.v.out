#! /c/Source/iverilog-install/bin/vvp
:ivl_version "12.0 (devel)" "(s20150603-1539-g2693dd32b)";
:ivl_delay_selection "TYPICAL";
:vpi_time_precision - 12;
:vpi_module "C:\iverilog\lib\ivl\system.vpi";
:vpi_module "C:\iverilog\lib\ivl\vhdl_sys.vpi";
:vpi_module "C:\iverilog\lib\ivl\vhdl_textio.vpi";
:vpi_module "C:\iverilog\lib\ivl\v2005_math.vpi";
:vpi_module "C:\iverilog\lib\ivl\va_math.vpi";
S_000001ee63a7adf0 .scope module, "full_adder_tb" "full_adder_tb" 2 4;
 .timescale -12 -12;
v000001ee63b183b0_0 .var "a_tb", 0 0;
v000001ee63b18270_0 .var "b_tb", 0 0;
v000001ee63b17cd0_0 .var "cin_tb", 0 0;
v000001ee63b18090_0 .net "cout_tb", 0 0, L_000001ee63b18e30;  1 drivers
v000001ee63b17870_0 .net "sum_tb", 0 0, L_000001ee63a92bd0;  1 drivers
S_000001ee63a7cef0 .scope module, "fa0" "full_adder" 2 12, 3 2 0, S_000001ee63a7adf0;
 .timescale -12 -12;
    .port_info 0 /INPUT 1 "a";
    .port_info 1 /INPUT 1 "b";
    .port_info 2 /INPUT 1 "cin";
    .port_info 3 /OUTPUT 1 "sum";
    .port_info 4 /OUTPUT 1 "cout";
L_000001ee63b18e30 .functor OR 1, L_000001ee63b184f0, L_000001ee63b17eb0, C4<0>, C4<0>;
v000001ee63a927d0_0 .net "a", 0 0, v000001ee63b183b0_0;  1 drivers
v000001ee63a92870_0 .net "b", 0 0, v000001ee63b18270_0;  1 drivers
v000001ee63a92910_0 .net "cin", 0 0, v000001ee63b17cd0_0;  1 drivers
v000001ee63ab7890_0 .net "cout", 0 0, L_000001ee63b18e30;  alias, 1 drivers
v000001ee63b17d70_0 .net "ha0_cout", 0 0, L_000001ee63b17eb0;  1 drivers
v000001ee63b17e10_0 .net "ha0_sum", 0 0, L_000001ee63ac67e0;  1 drivers
v000001ee63b17af0_0 .net "ha1_cout", 0 0, L_000001ee63b184f0;  1 drivers
v000001ee63b18130_0 .net "sum", 0 0, L_000001ee63a92bd0;  alias, 1 drivers
S_000001ee63a7d080 .scope module, "ha0" "half_adder" 3 13, 4 1 0, S_000001ee63a7cef0;
 .timescale -12 -12;
    .port_info 0 /INPUT 1 "a";
    .port_info 1 /INPUT 1 "b";
    .port_info 2 /OUTPUT 1 "sum";
    .port_info 3 /OUTPUT 1 "cout";
L_000001ee63ac67e0 .functor XOR 1, v000001ee63b183b0_0, v000001ee63b18270_0, C4<0>, C4<0>;
v000001ee63a92df0_0 .net "a", 0 0, v000001ee63b183b0_0;  alias, 1 drivers
v000001ee63ac5520_0 .net "b", 0 0, v000001ee63b18270_0;  alias, 1 drivers
v000001ee63a7d210_0 .net "cout", 0 0, L_000001ee63b17eb0;  alias, 1 drivers
v000001ee63a7d2b0_0 .net "sum", 0 0, L_000001ee63ac67e0;  alias, 1 drivers
L_000001ee63b17eb0 .arith/mult 1, v000001ee63b183b0_0, v000001ee63b18270_0;
S_000001ee63abb150 .scope module, "hf1" "half_adder" 3 22, 4 1 0, S_000001ee63a7cef0;
 .timescale -12 -12;
    .port_info 0 /INPUT 1 "a";
    .port_info 1 /INPUT 1 "b";
    .port_info 2 /OUTPUT 1 "sum";
    .port_info 3 /OUTPUT 1 "cout";
L_000001ee63a92bd0 .functor XOR 1, v000001ee63b17cd0_0, L_000001ee63ac67e0, C4<0>, C4<0>;
v000001ee63a92550_0 .net "a", 0 0, v000001ee63b17cd0_0;  alias, 1 drivers
v000001ee63a925f0_0 .net "b", 0 0, L_000001ee63ac67e0;  alias, 1 drivers
v000001ee63a92690_0 .net "cout", 0 0, L_000001ee63b184f0;  alias, 1 drivers
v000001ee63a92730_0 .net "sum", 0 0, L_000001ee63a92bd0;  alias, 1 drivers
L_000001ee63b184f0 .arith/mult 1, v000001ee63b17cd0_0, L_000001ee63ac67e0;
S_000001ee63ab7930 .scope task, "verify" "verify" 2 20, 2 20 0, S_000001ee63a7adf0;
 .timescale -12 -12;
v000001ee63b17b90_0 .var "a_task", 0 0;
v000001ee63b186d0_0 .var "b_task", 0 0;
v000001ee63b17c30_0 .var "cin_task", 0 0;
v000001ee63b18770_0 .var "cout_task", 0 0;
v000001ee63b17a50_0 .var "sum_task", 0 0;
TD_full_adder_tb.verify ;
    %load/vec4 v000001ee63b17b90_0;
    %store/vec4 v000001ee63b183b0_0, 0, 1;
    %load/vec4 v000001ee63b186d0_0;
    %store/vec4 v000001ee63b18270_0, 0, 1;
    %load/vec4 v000001ee63b17c30_0;
    %store/vec4 v000001ee63b17cd0_0, 0, 1;
    %delay 1, 0;
    %load/vec4 v000001ee63b17a50_0;
    %load/vec4 v000001ee63b17870_0;
    %cmp/e;
    %flag_get/vec4 4;
    %jmp/0 T_0.2, 4;
    %load/vec4 v000001ee63b18770_0;
    %load/vec4 v000001ee63b18090_0;
    %cmp/e;
    %flag_get/vec4 4;
    %and;
T_0.2;
    %flag_set/vec4 8;
    %jmp/0xz  T_0.0, 8;
    %vpi_call 2 33 "$display", "CORRECTO: Dato de entrada a: %b, Dato de entrada b: %b, Dato de entrada cin: %b, Dato de salida s esperado: %b, Dato de salida s obtenido: %b, Dato de salida c esperado: %b, Dato de salida c obtenido: %b", v000001ee63b183b0_0, v000001ee63b18270_0, v000001ee63b17c30_0, v000001ee63b17a50_0, v000001ee63b17870_0, v000001ee63b18770_0, v000001ee63b18090_0 {0 0 0};
    %jmp T_0.1;
T_0.0 ;
    %vpi_call 2 35 "$display", "INCORRECTO: Dato de entrada a: %b, Dato de entrada b: %b, Dato de entrada cin: %b, Dato de salida s esperado: %b, Dato de salida s obtenido: %b, Dato de salida c esperado: %b, Dato de salida c obtenido: %b", v000001ee63b183b0_0, v000001ee63b18270_0, v000001ee63b17c30_0, v000001ee63b17a50_0, v000001ee63b17870_0, v000001ee63b18770_0, v000001ee63b18090_0 {0 0 0};
T_0.1 ;
    %end;
    .scope S_000001ee63a7adf0;
T_1 ;
    %vpi_call 2 43 "$dumpfile", "full_adder_tb.vcd" {0 0 0};
    %vpi_call 2 44 "$dumpvars", 32'sb00000000000000000000000000000000, S_000001ee63a7adf0 {0 0 0};
    %end;
    .thread T_1;
    .scope S_000001ee63a7adf0;
T_2 ;
    %pushi/vec4 0, 0, 1;
    %store/vec4 v000001ee63b17b90_0, 0, 1;
    %pushi/vec4 0, 0, 1;
    %store/vec4 v000001ee63b186d0_0, 0, 1;
    %pushi/vec4 0, 0, 1;
    %store/vec4 v000001ee63b17c30_0, 0, 1;
    %pushi/vec4 0, 0, 1;
    %store/vec4 v000001ee63b17a50_0, 0, 1;
    %pushi/vec4 0, 0, 1;
    %store/vec4 v000001ee63b18770_0, 0, 1;
    %fork TD_full_adder_tb.verify, S_000001ee63ab7930;
    %join;
    %pushi/vec4 0, 0, 1;
    %store/vec4 v000001ee63b17b90_0, 0, 1;
    %pushi/vec4 0, 0, 1;
    %store/vec4 v000001ee63b186d0_0, 0, 1;
    %pushi/vec4 1, 0, 1;
    %store/vec4 v000001ee63b17c30_0, 0, 1;
    %pushi/vec4 1, 0, 1;
    %store/vec4 v000001ee63b17a50_0, 0, 1;
    %pushi/vec4 0, 0, 1;
    %store/vec4 v000001ee63b18770_0, 0, 1;
    %fork TD_full_adder_tb.verify, S_000001ee63ab7930;
    %join;
    %pushi/vec4 0, 0, 1;
    %store/vec4 v000001ee63b17b90_0, 0, 1;
    %pushi/vec4 1, 0, 1;
    %store/vec4 v000001ee63b186d0_0, 0, 1;
    %pushi/vec4 0, 0, 1;
    %store/vec4 v000001ee63b17c30_0, 0, 1;
    %pushi/vec4 1, 0, 1;
    %store/vec4 v000001ee63b17a50_0, 0, 1;
    %pushi/vec4 0, 0, 1;
    %store/vec4 v000001ee63b18770_0, 0, 1;
    %fork TD_full_adder_tb.verify, S_000001ee63ab7930;
    %join;
    %pushi/vec4 0, 0, 1;
    %store/vec4 v000001ee63b17b90_0, 0, 1;
    %pushi/vec4 1, 0, 1;
    %store/vec4 v000001ee63b186d0_0, 0, 1;
    %pushi/vec4 1, 0, 1;
    %store/vec4 v000001ee63b17c30_0, 0, 1;
    %pushi/vec4 0, 0, 1;
    %store/vec4 v000001ee63b17a50_0, 0, 1;
    %pushi/vec4 1, 0, 1;
    %store/vec4 v000001ee63b18770_0, 0, 1;
    %fork TD_full_adder_tb.verify, S_000001ee63ab7930;
    %join;
    %pushi/vec4 1, 0, 1;
    %store/vec4 v000001ee63b17b90_0, 0, 1;
    %pushi/vec4 0, 0, 1;
    %store/vec4 v000001ee63b186d0_0, 0, 1;
    %pushi/vec4 0, 0, 1;
    %store/vec4 v000001ee63b17c30_0, 0, 1;
    %pushi/vec4 1, 0, 1;
    %store/vec4 v000001ee63b17a50_0, 0, 1;
    %pushi/vec4 0, 0, 1;
    %store/vec4 v000001ee63b18770_0, 0, 1;
    %fork TD_full_adder_tb.verify, S_000001ee63ab7930;
    %join;
    %pushi/vec4 1, 0, 1;
    %store/vec4 v000001ee63b17b90_0, 0, 1;
    %pushi/vec4 0, 0, 1;
    %store/vec4 v000001ee63b186d0_0, 0, 1;
    %pushi/vec4 1, 0, 1;
    %store/vec4 v000001ee63b17c30_0, 0, 1;
    %pushi/vec4 0, 0, 1;
    %store/vec4 v000001ee63b17a50_0, 0, 1;
    %pushi/vec4 1, 0, 1;
    %store/vec4 v000001ee63b18770_0, 0, 1;
    %fork TD_full_adder_tb.verify, S_000001ee63ab7930;
    %join;
    %pushi/vec4 1, 0, 1;
    %store/vec4 v000001ee63b17b90_0, 0, 1;
    %pushi/vec4 1, 0, 1;
    %store/vec4 v000001ee63b186d0_0, 0, 1;
    %pushi/vec4 0, 0, 1;
    %store/vec4 v000001ee63b17c30_0, 0, 1;
    %pushi/vec4 0, 0, 1;
    %store/vec4 v000001ee63b17a50_0, 0, 1;
    %pushi/vec4 1, 0, 1;
    %store/vec4 v000001ee63b18770_0, 0, 1;
    %fork TD_full_adder_tb.verify, S_000001ee63ab7930;
    %join;
    %pushi/vec4 1, 0, 1;
    %store/vec4 v000001ee63b17b90_0, 0, 1;
    %pushi/vec4 1, 0, 1;
    %store/vec4 v000001ee63b186d0_0, 0, 1;
    %pushi/vec4 1, 0, 1;
    %store/vec4 v000001ee63b17c30_0, 0, 1;
    %pushi/vec4 1, 0, 1;
    %store/vec4 v000001ee63b17a50_0, 0, 1;
    %pushi/vec4 1, 0, 1;
    %store/vec4 v000001ee63b18770_0, 0, 1;
    %fork TD_full_adder_tb.verify, S_000001ee63ab7930;
    %join;
    %end;
    .thread T_2;
# The file index is used to find the file name in the following table.
:file_names 5;
    "N/A";
    "<interactive>";
    "full_adder_tb.v";
    "./../../src/verilog/full_adder.v";
    "./../../../Project 1 - Half Adder/src/verilog/half_adder.v";
